src/core/ddsi/src/ddsi_typelookup.c

```diff
@@ -0,0 +1,356 @@
+      memcpy (&tlm->endpoints[tlm->n_endpoints - 1], src_ep_guid, sizeof (*src_ep_guid));
+      GVTRACE (" add ep "PGUIDFMT, PGUID (*src_ep_guid));
+    }
+  }
+  if (tlm->sertype == NULL && type != NULL)
+  {
+    tlm->sertype = ddsi_sertype_ref (type);
+    tlm->state = TL_META_RESOLVED;
+    GVTRACE (" resolved");
+  }
+  if (dst != NULL && guid_prefix_empty (&tlm->dst_prefix) && !guid_prefix_empty (dst))
+    memcpy (&tlm->dst_prefix, dst, sizeof (*dst));
+  tlm->refc++;
+  GVTRACE (" state %d refc %u\n", tlm->state, tlm->refc);
+  ddsrt_mutex_unlock (&gv->tl_admin_lock);
+  if (type_id == NULL)
+    ddsrt_free (tid);
+}
+
+void ddsi_tl_meta_unref (struct ddsi_domaingv *gv, const type_identifier_t *type_id, const ddsi_guid_t *ep_guid)
+{
+  ddsrt_mutex_lock (&gv->tl_admin_lock);
+  struct tl_meta *tlm = ddsi_tl_meta_lookup_locked (gv, type_id);
+  assert (tlm != NULL);
+  ddsi_tl_meta_unref_impl_locked (gv, tlm, ep_guid);
+  ddsrt_mutex_unlock (&gv->tl_admin_lock);
+}
+
+static struct participant *get_typelookup_pp (const struct ddsi_domaingv *gv, const ddsi_guid_prefix_t *guid_prefix)
+{
+  struct entidx_enum_participant est;
+  struct participant *pp = NULL;
+  bool pp_found = false;
+  entidx_enum_participant_init (&est, gv->entity_index);
+  while (!pp_found && (pp = entidx_enum_participant_next (&est)) != NULL)
+  {
+    if (guid_prefix_empty (guid_prefix) || memcmp (&pp->e.guid.prefix, guid_prefix, sizeof (pp->e.guid.prefix)) == 0)
+      pp_found = true;
+  }
+  entidx_enum_participant_fini (&est);
+  return pp;
+}
+
+bool ddsi_tl_request_type (struct ddsi_domaingv * const gv, const type_identifier_t *type_id)
+{
+  ddsrt_mutex_lock (&gv->tl_admin_lock);
+  struct tl_meta *tlm = ddsi_tl_meta_lookup_locked (gv, type_id);
+  if (tlm->state != TL_META_NEW)
+  {
+    GVTRACE ("ddsi_tl_request_type - state not-new (%u) for "PTYPEIDFMT"\n", tlm->state, PTYPEID (*type_id));
+    ddsrt_mutex_unlock (&gv->tl_admin_lock);
+    return false;
+  }
+
+  struct participant *pp = get_typelookup_pp (gv, &tlm->dst_prefix);
+  struct writer *wr = get_builtin_writer (pp, NN_ENTITYID_TL_SVC_BUILTIN_REQUEST_WRITER);
+  if (wr == NULL)
+  {
+    GVTRACE ("ddsi_tl_request_type("PGUIDFMT") - builtin TypeLookup service request writer not found\n", PGUID (pp->e.guid));
+    ddsrt_mutex_unlock (&gv->tl_admin_lock);
+    return false;
+  }
+
+  type_lookup_request_t request;
+  memcpy (&request.writer_guid, &wr->e.guid, sizeof (wr->e.guid));
+  request.sequence_number = wr->seq;
+  request.type_ids.n = 1;
+  request.type_ids.type_ids = tlm->type_id;
+  struct ddsi_serdata *serdata = ddsi_serdata_from_sample (gv->tl_svc_request_type, SDK_DATA, &request);
+  serdata->timestamp = ddsrt_time_wallclock ();
+
+  GVTRACE (" wr-tl-req "PGUIDFMT" typeid "PTYPEIDFMT"\n", PGUID (wr->e.guid), PTYPEID(*type_id));
+  struct ddsi_tkmap_instance *tk = ddsi_tkmap_lookup_instance_ref (gv->m_tkmap, serdata);
+  write_sample_gc (lookup_thread_state (), NULL, wr, serdata, tk);
```

Comment:
I would prefer not holding any locks while writing this, or to use a guaranteed non-blocking write operation. That all depends on QoS and stuff like that — currently a `KEEP_ALL` writer isn't guaranteed non-blocking. But we could make it such for this case.